src/NHibernate/Dialect/MsSql2012Dialect.cs

```diff
@@ -58,25 +58,25 @@ protected override void RegisterFunctions()
 
 		public override SqlString GetLimitString(SqlString querySqlString, SqlString offset, SqlString limit)
 		{
+			if (offset == null)
+				return base.GetLimitString(querySqlString, null, limit);
+
 			using (var tokenEnum = new SqlTokenizer(querySqlString).GetEnumerator())
 			{
-				if (!tokenEnum.TryParseUntilFirstMsSqlSelectColumn()) return null;
+				if (!tokenEnum.TryParseUntilFirstMsSqlSelectColumn(out _, out var isDistinct)) return null;
+
+				var hasOrderBy = tokenEnum.TryParseUntil("order");
+				if(!hasOrderBy && isDistinct)
+						return base.GetLimitString(querySqlString, offset, limit);
 
 				var result = new SqlStringBuilder(querySqlString);
-				if (!tokenEnum.TryParseUntil("order"))
+				if (!hasOrderBy)
 				{
 					result.Add(" ORDER BY CURRENT_TIMESTAMP");
```

Comment:
Ordering by the first column of the result set seems less hackish to me than ordering by the current date and time. This thing attempts to emulate the behavior of unordered query on SQL-Server. I do not think NHibernate should try to do that, since NHibernate is more about abstracting away the specificity of database vendors rather than surfacing them.

I am tempted to say we should not even try to support unordered paged queries, because conceptually, writing such queries is a mistake: there is no predictability about which data will be sent back. There is not even any guarantee the paging will be consistent: it may include the same row in different pages, and not yield some rows whatever the asked page. (With SQL Server, it depends on the indexes used in the computed execution plan. If for some reasons it uses another plan, the ordering yielded by an unordered query will change.)
But removing all support for it will undoubtedly break a lot of existing application, so we have to keep that legacy.

So, replacing `CURRENT_TIMESTAMP` by `1` for this case looks better to me.

Using a fallback on the row_number paging seems to me as trying to somewhat preserve the default behavior of SQL-Server with unordered queries, which I do not think is valuable for NHibernate.

Reply:
> Ordering by the first column of the result set seems less hackish

IMO it's better to use existing behavior (hackish or not) than introduce another dialect specific hack (even less hackish) with different behavior.

Reply:
> I am tempted to say we should not even try to support unordered paged queries, because conceptually, writing such queries is a mistake

I fully agree. 

I think the user should see an error and then try to fix their query.